File: Week 1/cpp/1.installation.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: C++ Installation & IDE Setup
 * File: 1.installation.cpp
 *
 * CONCEPT:
 * Setting up a C++ toolchain. Unlike Java (which bundles JDK + JRE) C++
 * needs you to install a COMPILER (g++, clang++, MSVC) and optionally an
 * IDE/build system (VS Code + CMake, CLion, Visual Studio).
 *
 * KEY POINTS:
 *  - macOS:    `xcode-select --install`  (provides clang++)
 *  - Linux:    `sudo apt install build-essential`  (g++)
 *  - Windows:  Install MSYS2 + MinGW-w64, OR Visual Studio with C++ workload
 *  - Verify with: `g++ --version` or `clang++ --version`
 *  - Compile a single file:  g++ -std=c++17 -Wall -O2 hello.cpp -o hello
 *  - Run the produced binary: ./hello   (or hello.exe on Windows)
 *
 * SYNTAX:
 *  #include <iostream>           // standard I/O streams
 *  int main() { ... return 0; }   // entry point; returns int (0 = success)
 *
 * C++-SPECIFIC NOTES vs Java:
 *  - C++ is COMPILED to native code; Java compiles to JVM bytecode.
 *  - There is no `package` declaration; namespaces (`namespace foo {}`) play
 *    a similar role.
 *  - Header (.h/.hpp) files declare; source (.cpp) files define.
 *  - `main` is a free function — there is no enclosing class as in Java.
 *
 * DRY RUN:
 *  1. Install g++ -> `g++ --version` prints "g++ (Ubuntu 13.x.x)..."
 *  2. Save this file as 1.installation.cpp -> compile with
 *       g++ -std=c++17 1.installation.cpp -o setup
 *     and run `./setup` to print the environment summary below.
 *
 * COMPLEXITY: N/A (informational program).
 */

#include <iostream>
#include <vector>
#include <string>

int main() {
    std::cout << "=== C++ Environment Check ===\n";

#ifdef __clang__
    std::cout << "Compiler        : clang++ "
              << __clang_major__ << "." << __clang_minor__ << "\n";
#elif defined(__GNUC__)
    std::cout << "Compiler        : g++ "
              << __GNUC__ << "." << __GNUC_MINOR__ << "\n";
#elif defined(_MSC_VER)
    std::cout << "Compiler        : MSVC " << _MSC_VER << "\n";
#else
    std::cout << "Compiler        : unknown\n";
#endif

    std::cout << "C++ standard    : " << __cplusplus << "  (e.g. 201703 = C++17)\n";

    std::cout << "\n=== Setup Checklist ===\n";
    const std::vector<std::string> steps = {
        "1. Install a C++ compiler (g++ on Linux, clang++ on macOS, MSVC on Windows).",
        "2. Verify with: g++ --version  OR  clang++ --version",
        "3. Choose an IDE: VS Code + C/C++ extension, CLion, or Visual Studio.",
        "4. Optional: install CMake for cross-platform builds.",
        "5. Compile a hello.cpp:  g++ -std=c++17 -Wall hello.cpp -o hello",
        "6. Run:  ./hello",
    };
    for (const auto& step : steps) std::cout << step << '\n';

    return 0;
}

/*
 * NOTES:
 *  - Java is compile-once-run-anywhere on the JVM; C++ binaries are platform-
 *    specific (recompile per OS / architecture).
 *  - Java has automatic memory management (GC); C++ uses RAII + smart pointers.
 *  - Java's `System.out.println` -> C++'s `std::cout << ... << std::endl`.
 *  - C++ supports both procedural and OO styles; main() does not need a class.
 *  - Conditional compilation via `#ifdef __clang__` / `__GNUC__` / `_MSC_VER`
 *    is the C++ way to detect compiler — Java has no equivalent.
 */

File: Week 1/rust/1.installation.rs
/*
 * WEEK 1 - RUST FUNDAMENTALS
 * Topic: Rust Installation & Toolchain Setup
 * File: 1.installation.rs
 *
 * CONCEPT:
 * Rust is installed via `rustup`, the official toolchain manager. `rustup`
 * pulls down `rustc` (compiler), `cargo` (build tool + package manager),
 * `rustfmt` (formatter), and `clippy` (linter) in one step.
 *
 * KEY POINTS:
 *  - One-liner install: curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
 *  - Verify with: `rustc --version`, `cargo --version`
 *  - Update toolchain: `rustup update`
 *  - Create a project:  `cargo new hello`
 *  - Build:             `cargo build`
 *  - Run:               `cargo run`
 *  - Check (no codegen): `cargo check`   (fast feedback loop)
 *  - Test:              `cargo test`
 *  - Format:            `cargo fmt`
 *  - Lint:              `cargo clippy`
 *  - For a single file: `rustc 1.installation.rs && ./1.installation`
 *
 * SYNTAX:
 *  fn main() { ... }                  // entry point; takes no args
 *  println!("...")                     // macro (note the `!`)
 *
 * RUST-SPECIFIC NOTES vs Java:
 *  - Compiled to native code (like C++); no VM.
 *  - Memory safety enforced at COMPILE time via the borrow checker — no GC.
 *  - Macros end with `!` (e.g. `println!`, `vec!`).
 *  - Cargo handles dependencies (similar to Maven/Gradle for Java).
 *
 * DRY RUN:
 *  After `rustup install`, run `cargo new myapp && cd myapp && cargo run`
 *  -> prints "Hello, world!" using the auto-generated `src/main.rs`.
 *
 * COMPLEXITY: N/A.
 */

fn main() {
    println!("=== Rust Environment Check ===");
    // option_env! returns Option<&'static str> at compile time
    let pkg = option_env!("CARGO_PKG_NAME").unwrap_or("n/a (single-file build)");
    println!("cargo package: {}", pkg);
    println!("(run `rustc --version` to inspect compiler version)");

    println!();
    println!("=== Setup Checklist ===");
    let steps = [
        "1. Install rustup from https://rustup.rs/",
        "2. Verify: rustc --version  &&  cargo --version",
        "3. Add an editor/IDE: VS Code + rust-analyzer extension",
        "4. Create a project: cargo new myapp",
        "5. Run it:           cargo run",
        "6. Format / lint:    cargo fmt && cargo clippy",
        "7. Single-file mode: rustc file.rs && ./file",
    ];
    for s in steps.iter() {
        println!("{}", s);
    }
}

/*
 * NOTES:
 *  - Java needs JDK + IDE; Rust bundles the toolchain via rustup.
 *  - Java compiles to bytecode for the JVM; Rust compiles to a native binary.
 *  - Rust's Cargo is closest to Node's npm or Python's pip+setuptools — built in.
 *  - Macros (`println!`) look like functions but expand at compile time.
 *  - `env!()` reads compile-time env vars — used here just to demo macro use.
 */
